### ISSUE TITLE
sprite playfield collision bug

### ISSUE THREAD

**Issue by mithrendal** [dirkwhoffmann/vAmiga#416]
in gravity force both space ships are constantly exploding at their starting point... when sprite playfield col detection on ...

<img width="798" alt="image" src="https://user-images.githubusercontent.com/17108995/95041908-56915600-06d8-11eb-95a2-983eca719862.png">

when sprite playfield col detection off they stop exploding but can fly through the walls ..

[testcase_spriteplayfield_col.adf.zip](https://github.com/dirkwhoffmann/vAmiga/files/5325323/testcase_spriteplayfield_col.adf.zip)

tested on vAmiga0.9.12

**Comment by dirkwhoffmann** [comment]
I can reproduce half of it. If collision detection is enabled, the space ship crashes immediately. If collision detection is off, I still crash into the surrounding. By "wall", do you mean the rocks or whatever the brown stuff is?

**Comment by dirkwhoffmann** [comment]
Findings: Game initializes CLXCON with $F041.

<img width="478" alt="Bildschirmfoto 2020-10-10 um 00 26 59" src="https://user-images.githubusercontent.com/12561945/95636400-56ee7000-0a8f-11eb-8aee-ffbd4920a597.png">

This means: 

- All sprites are subject to collision testing.
- A collision is detected if and only if a non-transparent sprite pixel overlaps with a pixels having a 1 in bitplane 1.

Therefore, only playfield 1 plays a role in collision detection. In other words: None of the playfield 2 bits are taken into account. Hence, I think this already exposes the bug. This is my code for collision testing with playfield 2 pixels: 

 ```
        // Check for a collision with playfield 2
        if ((bBuffer[pos] & enabled2) == compare2) {
            debug(CLX_DEBUG, "S%d collides with PF2\n", x);
            SET_BIT(clxdat, 5 + (x / 2));
            SET_BIT(clxdat, 1 + (x / 2));
        }
```

If the enable mask equals 0 (which is the case in this game), the condition becomes true whenever `compare2` equals 0. However, the condition is supposed to be false no matter what. The correct equation is: 

```
        // Check for a collision with playfield 2
        if (enabled2 && (bBuffer[pos] & enabled2) == compare2) {
            debug(CLX_DEBUG, "S%d collides with PF2\n", x);
            SET_BIT(clxdat, 5 + (x / 2));
            SET_BIT(clxdat, 1 + (x / 2));
        }
```

Using this modification, the game seems to work fine.

TODOs before closing: 
- Add vAmigaTS test exposing this bug.
- Run all existing tests in vAmigaTS/Sprites/collision.

Very nice game, btw 😎.

**Comment by mithrendal** [comment]
>If collision detection is off, I still crash into the surrounding. By "wall", do you mean the rocks or whatever the brown stuff is?

thats true they do a boundary check on the frames I think ... but you wont crash into the bezels its more of a bounce back (gummiwand).  But you are supposed to crash ...

here is a picture col detect off two player mission ... I manoevered the gray ship slightly right to the [8]... it hides behind the playfield the brown earth and the green basement... can you see it ?  
<img width="794" alt="image" src="https://user-images.githubusercontent.com/17108995/95647467-f7648480-0acf-11eb-82d1-c31d27efd02c.png">

**Comment by mithrendal** [comment]
>` if (enabled2 && (bBuffer[pos] & enabled2) == compare2) `

yes that looks good ... it is faster too 🚀

**Comment by dirkwhoffmann** [comment]
> yes that looks good ... it is faster .. I mean energy efficient too 🚀

Unfortunately, it's still wrong. 🙄
The HRM states: "If all bitplanes are excluded (disabled), then a bitplane collision will always be detected."

**Comment by dirkwhoffmann** [comment]
I think it was something very stupid. In the original code

```
       // Check for a collision with playfield 2
       if ((bBuffer[pos] & enabled2) == compare2) {
           debug(CLX_DEBUG, "S%d collides with PF2\n", x);
           SET_BIT(clxdat, 5 + (x / 2));
           SET_BIT(clxdat, 1 + (x / 2));  <---
       }
```

the marked line simply doesn't make sense. After removing it, all 16 new sprite collision tests pass and the game works, too.

With tests spritecoll8 and spritecoll8d, there are two new tests that verify the famous collision logic bug. If no collision with playfield 2 is detected in singe-playfield mode, playfield 1 won't signal a collision either, no matter what the pixel values say. Dual-playfield mode is not affected by this hardware bug.

**Comment by dirkwhoffmann** [comment]
Fixed in v0.9.13
